### PR TITLE
Repeated fields validation

### PR DESCRIPTION
**Pull request by dmdashenkov**
From now, the custom `Message` field options will be applied only to those `repeated` fields which are declared as `valid`.

#### Example
```protobuf
message Customer {
    // The validation is applied to all the elements of the repeated field.
    repeated string contact_name = 1 [(valid) = true, (pattern).regex = ".+\\s+.+", (required) = true];
   
    // The validation is applied to the values of the map field.
    map<string, people.PersonName> position_in_company = 2 [(valid) = true];

    // The validation is applied to the signular fields anyway.
    google.ptotobuf.Timestamp established = 3 [(when).in = PAST];
}
```

Note, that if a `repeated` or `map` field is declared as `required` (e.g. `Customer.contact_name`), the elements of it may be _default_ by themselves. For example `customerVBuilder.addContactName("")` is not a constaint violation.

#### Also in this PR:

 - the repository code style settings are synchronized with the `core-java` ones;
 - a few validation bugs are fixed;
 - some tests for the `FieldValidator`s and `ValidatingBuilder`s are added.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: base/src/test/java/io/spine/validate/FieldValidatorShould.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import static io.spine.test.Verify.assertNotEmpty;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * @author Dmytro Dashenkov
+ */
+public abstract class FieldValidatorShould<V> {
+
+    protected abstract FieldValidator<V>
+    getValidatorForValidatedRequiredField(ImmutableList<V> values);
```
Comment (armiol) [inline base/src/test/java/io/spine/validate/FieldValidatorShould.java]: Not sure why you need the whole `ImmutableList<V>`, when `Class<V>` seems to be enough.

File: base/src/test/proto/spine/test/validate/msg/validating_builder_test.proto [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+    string text = 3 [(required) = true];
+}
+
+message Signature {
+
+    bytes value = 1 [(required) = true];
+
+    string name = 2 [(pattern).regex = ".+\\s+.+"];
+
```
Comment (armiol) [inline base/src/test/proto/spine/test/validate/msg/validating_builder_test.proto]: Please remove this line.

File: ext.gradle [hunk 3]
```diff
@@ -41,7 +41,7 @@ ext {
 
     // We use Spine tools and model compiler in the build process.
     spineToolsVersion = '0.9.25-SNAPSHOT'
-    spineModelCompilerVersion = '0.9.34-SNAPSHOT'
+    spineModelCompilerVersion = '0.9.36-SNAPSHOT'
```
Comment (armiol) [inline ext.gradle]: This should probably be `0.9.35-SNAPSHOT`.

File: base/src/main/java/io/spine/validate/ByteStringFieldValidator.java [hunk 4]
```diff
@@ -57,11 +55,14 @@ protected boolean isValueNotSet(ByteString value) {
...
+     *
+     * <p>Performs no validation.
+     */
     @Override
-    protected List<ConstraintViolation> validate() {
-        checkIfRequiredAndNotSet();
-        final List<ConstraintViolation> violations = super.validate();
-        return violations;
+    protected void doValidate() {
+        // NoOp
```
Comment (armiol) [inline base/src/main/java/io/spine/validate/ByteStringFieldValidator.java]: Requires some explanation in JavaDocs. I am not sure that's correct, as there may be some rules applied to it. Not right now, but in general — it's always a good idea to ask `super`.

Same below.

Comment (dmdashenkov) [inline base/src/main/java/io/spine/validate/ByteStringFieldValidator.java]: Done.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

File: .idea/inspectionProfiles/Project_Default.xml [hunk 5]
```diff
@@ -148,6 +148,11 @@
     <inspection_tool class="ConfusingMainMethod" enabled="true" level="WARNING" enabled_by_default="true" />
     <inspection_tool class="ConfusingOctalEscape" enabled="true" level="WARNING" enabled_by_default="true" />
     <inspection_tool class="ConstantAssertCondition" enabled="true" level="WARNING" enabled_by_default="true" />
+    <inspection_tool class="ConstantConditions" enabled="true" level="WARNING" enabled_by_default="true">
```
Comment (armiol) [inline .idea/inspectionProfiles/Project_Default.xml]: We have just updated the settings from `core-java` today. I don't think this file should be modified now.

File: base/src/main/java/io/spine/validate/FieldValidator.java [hunk 6]
```diff
@@ -252,6 +300,11 @@ private boolean isRequiredEntityIdField() {
         return result;
     }
 
+    protected boolean isRepeatedOrMap() {
+        return fieldDescriptor.isRepeated() || fieldDescriptor.isMapField();
```
Comment (armiol) [inline base/src/main/java/io/spine/validate/FieldValidator.java]: This line is reported to be only partially tested.

File: base/src/main/java/io/spine/validate/StringFieldValidator.java [hunk 7]
```diff
@@ -57,11 +55,8 @@
     }
 
     @Override
-    protected List<ConstraintViolation> validate() {
-        checkIfRequiredAndNotSet();
+    protected void doValidate() {
         checkIfMatchesToRegexp();
```
Comment (armiol) [inline base/src/main/java/io/spine/validate/StringFieldValidator.java]: Let's rename to `checkMatchesRegexp`.

File: base/src/test/java/io/spine/validate/FieldValidatorShould.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+public abstract class FieldValidatorShould<V> {
+
+    protected abstract FieldValidator<V> validatedRequiredRepeatedField(List<V> values);
+
+    protected abstract FieldValidator<V> requiredRepeatedField(List<V> values);
+
+    protected abstract FieldValidator<V> validatedRepeatedField(List<V> values);
+
+    protected abstract FieldValidator<V> uncheckedRepeatedField(List<V> values);
+
```
Comment (armiol) [inline base/src/test/java/io/spine/validate/FieldValidatorShould.java]: Please remove this line.

File: base/src/test/java/io/spine/validate/FieldValidatorShould.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+import static com.google.common.collect.ImmutableList.of;
+import static io.spine.test.Verify.assertNotEmpty;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+/**
+ * @author Dmytro Dashenkov
+ */
+public abstract class FieldValidatorShould<V> {
+
+    protected abstract FieldValidator<V> validatedRequiredRepeatedField(List<V> values);
```
Comment (armiol) [inline base/src/test/java/io/spine/validate/FieldValidatorShould.java]: As per my previous PR review. Something does not feel right, as you are passing the list, but obtaining a validator for the list item. From the signature I don't understand, why `Class<V> value` is not enough as an argument.

From the implementation I understand there is a FieldValidator, which requires something to validate. And you pass the list of values to validate. But that's totally unclear from both the method name and the absence of Javadoc. Perhaps, should be both renamed and documented.

File: base/src/test/proto/spine/test/validate/msg/messages.proto [hunk 10]
```diff
@@ -172,3 +173,45 @@ message EnclosedMessageFieldValueWithCustomInvalidMessage {
...
+message MessageWithRepeatedRequiredValidatedMessageField {
+    repeated InvalidMessage bound = 1 [(valid) = true, (required) = true];
+}
+
+// A message which is invalid by design.
+//
+// The field `s` must contain a non-empty value, but it also must match "^$" regex, which represents
+// an empty string. Thus, any instance of this message type is invalid.
+message InvalidMessage {
+    string s = 1 [(required) = true, (pattern).regex = "^$"];
```
Comment (armiol) [inline base/src/test/proto/spine/test/validate/msg/messages.proto]: Please avoid `s` as a variable name.

File: base/src/test/proto/spine/test/validate/msg/validating_builder_test.proto [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    // This field is both required and validated.
+    //
+    repeated string name = 1 [(valid) = true, (pattern).regex = ".+\\s+.+", (required) = true];
+
+    // The contact emails.
+    //
+    // This field is validated but not required.
+    repeated string email = 2 [(valid) = true, (pattern).regex = ".+@.+\\..+"];
+
+    // The map of the positions in the company to the peaple who occupy them.
```
Comment (armiol) [inline base/src/test/proto/spine/test/validate/msg/validating_builder_test.proto]: `peaple`

File: base/src/test/proto/spine/test/validate/msg/validating_builder_test.proto [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    //
+    // This field is required but never validated.
```
Comment (armiol) [inline base/src/test/proto/spine/test/validate/msg/validating_builder_test.proto]: `This field is required, but the contents of the messages are never validated.`

File: base/src/test/proto/spine/test/validate/msg/validating_builder_test.proto [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    //
+    // This field is neither required, nor validated.
+    //
+    map<string, Contract> old_contract = 6;
+
+    // The signatures of the company proxy people.
+    //
+    // This field is required but not validated.
+    //
+    repeated Signature sign = 7 [(required) = true];
```
Comment (armiol) [inline base/src/test/proto/spine/test/validate/msg/validating_builder_test.proto]: `repeated Signature signature`

File: base/src/test/proto/spine/test/validate/msg/validating_builder_test.proto [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+option java_package = "io.spine.test.validate.msg.builder";
+
+import "google/protobuf/timestamp.proto";
+
+import "spine/options.proto";
+
+import "spine/net/email_address.proto";
+import "spine/people/person_name.proto";
+
+message Customer {
```
Comment (armiol) [inline base/src/test/proto/spine/test/validate/msg/validating_builder_test.proto]: Judging on its contents, this isn't a `Customer`. All the repeated fields look weird in this context. Please rethink the example. I would use `Project` with attachments, tasks, members. Some of those could have nested collections; e.g. a task should probably have attachments (as well as a project), comments etc. 

`attachment` can be a map (MIME type string -> `Attachment`). And if you need another map, you could have something like `attributes`, like we have for a `Command` message.

File: base/src/test/proto/spine/test/validate/msg/validating_builder_test.proto [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+    //
+    // This field validated but not required.
+    //
+    map<string, Contract> contract = 5 [(valid) = true, (required) = false];
+
+    // The old contracts with this customer.
+    //
+    // This field is neither required, nor validated.
+    //
+    map<string, Contract> old_contract = 6;
```
Comment (armiol) [inline base/src/test/proto/spine/test/validate/msg/validating_builder_test.proto]: Perhaps, these are `expired_contracts`. In any case, those are values. And it's totally unclear, what the keys are.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

File: base/src/main/java/io/spine/validate/FieldValidatorFactory.java [hunk 16]
```diff
@@ -33,29 +43,63 @@
...
     /**
      * Creates a new validator instance according to the field type and validates the field.
      *
+     * <p>The target field of the resulting validator is represented with a linear data structure,
+     * i.e. not a map.
+     *
+     * @param descriptor    a descriptor of the field to validate
+     * @param fieldValue    a value of the field to validate
+     * @param rootFieldPath a path to the root field
+     * @param strict        if {@code true} validators would always assume that the field is
```
Comment (armiol) [inline base/src/main/java/io/spine/validate/FieldValidatorFactory.java]: This sentence looks incomplete.

File: base/src/test/java/io/spine/validate/FieldValidatorShould.java [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+/**
+ * @author Dmytro Dashenkov
+ */
+public abstract class FieldValidatorShould<V> {
+
+    /**
+     * Creates a {@link FieldValidator} for a repeated required validated field.
+     *
+     * <p>The proto description of this field looks as follows:
+     * {repeated T field_name = 42 [(required) = true, (valid) = true];}, where {@code T} is
```
Comment (armiol) [inline base/src/test/java/io/spine/validate/FieldValidatorShould.java]: `{@code ...}` looks missing. Same below.

File: base/src/test/java/io/spine/validate/FieldValidatorShould.java [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+     * the type parameter of the tested {@link FieldValidator}.
+     *
+     * @param values the list of the field values
+     * @return new instance of {@link FieldValidator}
+     */
+    protected abstract FieldValidator<V> validatedRequiredRepeatedFieldValidator(List<V> values);
+
+    /**
+     * Creates a {@link FieldValidator} for a repeated required but not validated field.
```
Comment (armiol) [inline base/src/test/java/io/spine/validate/FieldValidatorShould.java]: `for a repeated required field, which child items are not validated separately.`

File: base/src/test/java/io/spine/validate/FieldValidatorShould.java [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * @author Dmytro Dashenkov
+ */
+public abstract class FieldValidatorShould<V> {
+
+    /**
+     * Creates a {@link FieldValidator} for a repeated required validated field.
```
Comment (armiol) [inline base/src/test/java/io/spine/validate/FieldValidatorShould.java]: `...for a repeated required field, which child items are also validated separately.`

File: base/src/test/java/io/spine/validate/FieldValidatorShould.java [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+    protected abstract FieldValidator<V> uncheckedRepeatedFieldValidator(List<V> values);
+
+    /**
+     * Creates a {@link FieldValidator} for a map field.
+     *
+     * <p>The resulting validator should be empty (i.e. contain no values).
+     *
+     * @return new instance of {@link FieldValidator}
+     */
+    protected abstract FieldValidator<V> mapFieldValidator();
```
Comment (armiol) [inline base/src/test/java/io/spine/validate/FieldValidatorShould.java]: Let's call it `emptyMapFieldValidator` then.

File: base/src/test/proto/spine/test/validate/msg/messages.proto [hunk 21]
```diff
@@ -172,3 +173,77 @@ message EnclosedMessageFieldValueWithCustomInvalidMessage {
...
+
+message MessageWithMapFloatField {
+    map<string, float> map_field = 1;
+}
+
+message MessageWithMapDoubleField {
+    map<string, double> map_field = 1;
+}
+
+// A message which is invalid by design.
```
Comment (armiol) [inline base/src/test/proto/spine/test/validate/msg/messages.proto]:
```
A message which is used to violate its constraints in scope of `MessageFieldValidator` tests.
```

File: base/src/test/proto/spine/test/validate/msg/messages.proto [hunk 22]
```diff
@@ -172,3 +173,77 @@ message EnclosedMessageFieldValueWithCustomInvalidMessage {
...
+//
+// The field `s` must contain a non-empty value, but it also must match "^$" regex, which represents
```
Comment (armiol) [inline base/src/test/proto/spine/test/validate/msg/messages.proto]:
```
The field  `invalid_field` ...
```

File: base/src/test/proto/spine/test/validate/msg/validating_builder_test.proto [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+
+import "google/protobuf/timestamp.proto";
+
+import "spine/options.proto";
+
+import "spine/net/email_address.proto";
+import "spine/net/url.proto";
+import "spine/people/person_name.proto";
+
+// The test definitions for the ValidatingBuilder tests.
```
Comment (armiol) [inline base/src/test/proto/spine/test/validate/msg/validating_builder_test.proto]:
```
for the `ValidatingBuilder` tests.
```

File: base/src/test/proto/spine/test/validate/msg/validating_builder_test.proto [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+//
+// Use in ValidatingBuilderShould.java only.
```
Comment (armiol) [inline base/src/test/proto/spine/test/validate/msg/validating_builder_test.proto]: `Use in `ValidatingBuilderShould` only.`

File: base/src/main/java/io/spine/validate/FieldValidatorFactory.java (hunk 16 above)
Comment (dmdashenkov) [inline base/src/main/java/io/spine/validate/FieldValidatorFactory.java]: Done

**Pull request merged**